tasty-query/shared/src/main/scala/tastyquery/Symbols.scala

```diff
@@ -417,6 +417,78 @@ object Symbols {
+      *   `runtimeClass.isSubclass(this.owner.asClass)` is false
+      * @throws tastyquery.Exceptions.MemberNotFoundException
+      *   if no concrete matching symbol could be found
+      */
+    final def runtimeImplementingSymbol(runtimeClass: ClassSymbol)(using Context): TermSymbol =
+      if canMatchInheritedSymbols then
+        val linearization = runtimeClass.linearization
+        if !linearization.contains(this.owner.asClass) then
+          throw IllegalArgumentException(s"Invalid run-time class $runtimeClass for dynamic super call to $this")
+        findRuntimeSymbolCommon(runtimeClass, linearization)
+      else
+        if !this.owner.isClass then
+          throw IllegalArgumentException(s"Nonsensical run-time implementation lookup for local symbol $this")
+        this
+    end runtimeImplementingSymbol
+
+    /** Resolves a dynamically bound `super` call for an object of class `runtimeClass`.
+      *
+      * This method returns the next symbol for a `super` call of the form
+      * `enclosingClass.super.foo` for an object of class `runtimeClass`.
+      * That is the first non-abstract symbol matching this symbol in the
+      * linearization of `runtimeClass`, after the occurrence of
+      * `enclosingClass`.
+      *
+      * @throws java.lang.IllegalArgumentException
+      *   if any of the following conditions happen:
+      *   `this.owner.isClass` is false;
+      *   or this symbol is private;
+      *   or this symbol is a constructor;
+      *   or `runtimeClass.isSubclass(this.owner.asClass)` is false;
+      *   or if `runtimeClass.isSubclass(enclosingClass)` is false
+      * @throws tastyquery.Exceptions.MemberNotFoundException
+      *   if no concrete matching symbol could be found
+      */
+    final def nextSuperSymbol(runtimeClass: ClassSymbol, enclosingClass: ClassSymbol)(using Context): TermSymbol =
+      if !canMatchInheritedSymbols then
+        throw IllegalArgumentException(s"Invalid dynamic super call to statically bound symbol $this")
+      val linearization = runtimeClass.linearization
+      if !linearization.contains(this.owner.asClass) then
+        throw IllegalArgumentException(s"Invalid run-time class $runtimeClass for dynamic super call to $this")
+      val suffix = linearization.dropWhile(_ != enclosingClass)
+      if suffix.isEmpty then
+        throw IllegalArgumentException(s"Invalid enclosing class $enclosingClass for dynamic super call to $this")
+      findRuntimeSymbolCommon(runtimeClass, suffix.tail)
+    end nextSuperSymbol
+
+    /** Common loop for `runtimeImplementingSymbol` and `nextSuperSymbol`. */
+    @tailrec
+    private def findRuntimeSymbolCommon(runtimeClass: ClassSymbol, linearization: List[ClassSymbol])(
+      using Context
+    ): TermSymbol =
+      linearization match
+        case inClass :: linearizationTail =>
+          matchingDecl(inClass, siteClass = runtimeClass) match
+            case Some(matchingSym) if !matchingSym.is(Abstract) =>
```

Comment:
I'm still a little skeptical of filtering out all abstract matching symbols here. But I cannot think of a case in which you need to find an abstract method when resolving a method call at runtime. With super calls inside traits, it doesn't happen because they are replaced with super _accessor_ calls, which although abstract, you never need to actually find, because you can find their referent just from the name. Maybe some edge case with abstract classes?

Reply:
If we did *not* exclude abstract methods, what would you then do? You can't actually call an abstract method, since there is no body.

Reply:
Fair… like I said, I still can’t think of a case in which it would be necessary, so this is probably correct as is. I mentioned abstract classes because that is something we didn’t talk about when we were discussing these methods.